[debops.cryptsetup] Fixes present state flag behavior

inthehack opened the pull request:
The state flag of the plaintext block devices controls if the block
devices must be opened or not and if the filesystem on it must be
mounted.

In case the filesystem is not managed by the role and the state flag is
set to `present', the plaintext block device should not be closed as it may be
used by other roles like debops.lvm.

Then, this commit adds an additional condition inside the teardown procedure
at the end of the role in order to implemente this use case.

In a further patch, the lvm-based use case should be better covered with
explicit configuration other than simply disable filesystem management.

Commit:
[debops.cryptsetup] Fixes present state flag behavior

The state flag of the plaintext block devices controls if the block
devices must be opened or not and if the filesystem on it must be
mounted.

In case the filesystem is not managed by the role and the state flag is
set to `present', the plaintext block device should not be closed as it may be
used by other roles like debops.lvm.

Then, this commit adds an additional condition inside the teardown procedure
at the end of the role in order to implemente this use case.

In a further patch, the lvm-based use case should be better covered with
explicit configuration other than simply disable filesystem management.

drybjed reviewed (dismissed):
Sounds reasonable to me. I'll wait for @ypid a day or two before merging, in case he sees any issues.

ypid commented:
Sounds also reasonable to me, thanks @jpeeters. I would like to check and test this in more detail. @drybjed Please wait with merging until Tuesday. Until then I should be able review it more.

ypid reviewed (changes requested):
This change is not in line with the docs of the role. Refer to https://docs.debops.org/en/master/ansible/roles/debops.cryptsetup/defaults-detailed.html#cryptsetup-devices-state-present So this would be a breaking change. I would suggest to introduce a new state called "mapped" instead which does what you want. Please update the docs as well ;-)

inthehack commented:
Ok I got i @ypid. Thx for the review. I will deal with this next week.

Commit:
[debops.cryptsetup] Fixes consistency of device mapping state

This commit fixes the consistency of device mapping state according to
previous patch suggestions. It adds a new state named `mapped` that
behaves like `mounted` but without mounting a file system. The behavior
of the `present` state is left untouched then.

This commit offers consistent working with LVM logical volumes. However,
it does not deal with controller managed encrypted devices, which would
require more logic for setting up LVM after opening the encrypted device
on boot.

inthehack commented:
@ypid @drybjed I pushed a new commit that should implement proposed suggestions from @ypid. Let me know if this requires more improvements. This docs should be up to date too.

ypid commented:
@jpeeters Thanks for the update. I would like to review and test this a bit. Will probably take a few weeks.

ypid commented on ansible/roles/debops.cryptsetup/tasks/manage_devices.yml:
```diff
@@ -269,7 +269,8 @@
          (item.0.create_filesystem|d(item.0.manage_filesystem|d(True)) | bool) and
-         not (item.0.swap|d(False) | bool))
+         not (item.0.swap|d(False) | bool) and
+         not (item.state|d(cryptsetup__state) in ['mapped'])
```
Missing `)`.

ypid reviewed (changes requested):
Thanks for the update. I have an issue with this patch because it does not "compile" so I assume it is not so well tested. After I fixed the "compile error", the "Manage fstab and mount state of the plaintext device mapper targets" task fails. This task also needs to be excluded from the "mapped" state. I stopped my review of the tasks file at this point.

Regarding the docs, I know they are quite extensive and not easy to change. It would be great if you could read defaults-detailed.rst as a whole and see if any other options of the role might be affected by the new state. At a minimum the difference between manage_filesystem and the mapped state should be mentioned. Bonus points for a full example showing an actual use case for the `mapped` state. This role is quite complex and I feel it needs easy and practical examples for it to be usable.
